### PR TITLE
Create neural network in OOP C#

### PR DESCRIPTION
**Pull request by AnneEjsing**

### CODE REVIEW

File: benchmarks/NN/oop_c#/Program.cs [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+
+
+        }
+    }
+
+    public static class Utils
+    {
+        public static double[,] LoadCSV(string filepath)
+        {
+            Random rnd = new Random();
```
Comment (Caspersn2) [inline benchmarks/NN/oop_c#/Program.cs:39]:
```suggestion
```

Bliver rnd brugt?

File: benchmarks/NN/oop_c#/Program.cs [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+                if (actual[i] == predictions[i])
+                    correct++;
+            return ((double)correct / numPredictions) * 100;
+        }
+    }
+    public static class MatrixExtensions
+    {
+        /// <summary>
+        /// Returns the row with number 'row' of this matrix as a 1D-Array.
+        /// </summary>
```
Comment (Caspersn2) [inline benchmarks/NN/oop_c#/Program.cs:351]: Elsker at der lige kommer summary på her, men det ikke er der nogen andre steder :P

**Comment by Caspersn2** [comment]
Vores data ligger i filer der ikke hedder det samme :P

File: benchmarks/NN/oop_c#/Program.cs (hunk 2 above)
Comment (AnneEjsing) [reply benchmarks/NN/oop_c#/Program.cs:351]: Det er skønheden ved copy/paste kode. xD

**Review by Caspersn2 — APPROVED** [review]

File: benchmarks/NN/oop_c#/Program.cs [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+                    train.SetRow(trainindex++, dataset.GetRow(i));
+            }
+            return (test, train);
+        }
+
+        private static T[,] shuffle<T>(Random random, T[,] array)
+        {
+            T[,] shuffledArray = new T[array.GetLength(0), array.GetLength(1)];
+            int numsRows = array.GetLength(0);
+            var indicies = Enumerable.Range(0, numsRows - 1).OrderBy(a => Guid.NewGuid()).ToArray();
```
Comment (Caspersn2) [inline benchmarks/NN/oop_c#/Program.cs:95]: Altså, altså, altså...

Comment (AnneEjsing) [reply benchmarks/NN/oop_c#/Program.cs]: Fixed it

**Comment by lrecht** [comment]
Hvorfor bliver wheat-seeds.csv ændret?

File: benchmarks/NN/oop_c#/Program.cs [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            {
+                if (i < numTest)
+                    test.SetRow(testindex++, dataset.GetRow(i));
+                else
+                    train.SetRow(trainindex++, dataset.GetRow(i));
+            }
+            return (test, train);
+        }
+
+        private static T[,] shuffle<T>(Random random, T[,] array)
```
Comment (lrecht) [inline benchmarks/NN/oop_c#/Program.cs:88]:
```suggestion
        private static T[,] shuffle<T>(T[,] array)

```

File: benchmarks/NN/oop_c#/Program.cs [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        {
+            T[,] shuffledArray = new T[array.GetLength(0), array.GetLength(1)];
+            int numRows = array.GetLength(0);
+            int[] indicies = new int[numRows];
+            Random rnd = new Random();
```
Comment (lrecht) [inline benchmarks/NN/oop_c#/Program.cs:93]:
```suggestion
            Random rnd = new Random(2);

```
Plejer vi ikke at have seed på vores random, så vi er sikre på det er det samme vi udregner i alle benchmarks?

Comment (lrecht) [reply benchmarks/NN/oop_c#/Program.cs:93]: På den anden side kan jeg jo også se at jeg sortere på en anden måde uanset...

File: benchmarks/NN/oop_c#/Program.cs [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+            foreach (Layer layer in layers)
+                layer.UpdateWeights(row, learning_rate);
+        }
+
+        private double computeError(double sumError, double[] outputs, double[] expected)
+        {
+            for (int i = 0; i < nOutputs; i++)
+                sumError += Math.Pow(expected[i] - outputs[i], 2);
+            return sumError;
+        }
```
Comment (lrecht) [inline benchmarks/NN/oop_c#/Program.cs:237]:
```suggestion

```
Så vidt jeg kan se bliver denne ikke kaldt..?

File: benchmarks/NN/oop_c#/Program.cs [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+                    train.SetRow(trainindex++, dataset.GetRow(i));
+            }
+            return (test, train);
+        }
+
+        private static T[,] shuffle<T>(Random random, T[,] array)
+        {
+            T[,] shuffledArray = new T[array.GetLength(0), array.GetLength(1)];
+            int numRows = array.GetLength(0);
+            int[] indicies = new int[numRows];
```
Comment (lrecht) [inline benchmarks/NN/oop_c#/Program.cs:92]:
```suggestion

```

File: benchmarks/NN/oop_c#/Program.cs [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+                sumError += Math.Pow(expected[i] - outputs[i], 2);
+            return sumError;
+        }
+    }
+
+    public class Layer
+    {
+        public Neuron[] Neurons { get; set; }
+        int numNeurons { get; set; }
+        IActivationStrategy activation { get; set; }
```
Comment (lrecht) [inline benchmarks/NN/oop_c#/Program.cs:244]: Man kunne vel argumentere for at det er neuronens ansvar at aktivere.

**Review by lrecht — APPROVED** [review]

File: benchmarks/NN/oop_c#/Program.cs [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+                double diff = max - min;
+                for (int j = 0; j < column.Length; j++)
+                    dataset[j, i] = (dataset[j, i] - min) / diff;
+            }
+            return dataset;
+        }
+
+        public static (double[,], double[,]) GetTestTrainSplit(double[,] dataset, double percent)
+        {
+            dataset = shuffle(new Random(), dataset);
```
Comment (lrecht) [inline benchmarks/NN/oop_c#/Program.cs]:
```suggestion
            dataset = shuffle(dataset);

```
